### PR TITLE
mempool: send events in a separate goroutine

### PR DESCRIPTION
**Pull request by fyrchik**
Related #2063.

Here is a scenario for a deadlock:
1. Notary pool is listening for events and possibly forms new transaction (`OnNewRequest`) which calls `Blockchain.PoolTx` in callback.
2. `AddBlock` removes stale transactions in mempool, and generates events on the way. As comment indicates, `RemoveStale` should run with blockchain mutex taken.

If `PoolTx` in (1) is called during (2), there will be a deadlock. Because `PoolTx` needs to take blockchain mutex on read.

Current solution: send events in a separate goroutine, but I am not yet convinced this fixes the problem completely, will do some more investigation.

Signed-off-by: Evgeniy Stratonikov <evgeniy@nspcc.ru>

### CODE REVIEW

**Comment by roman-khimov** [comment]
I'd rather create a separate goroutine to send transactions from Notary module, let it accept complete transactions and do `n.onTransaction()`. Add some buffer and then just drop transactions if it doesn't fit (they'll be retried anyway with the next block).

**Review by roman-khimov — APPROVED** [review]
Looks legit. @AnnaShaleva?

**Comment by roman-khimov** [comment]
Tests are not perfect though:
```
==================
WARNING: DATA RACE
Read at 0x00c000298071 by goroutine 85:
  github.com/nspcc-dev/neo-go/pkg/core.TestNotary.func1()
      /go/src/github.com/nspcc-dev/neo-go/pkg/core/notary_test.go:76 +0xfd
  github.com/nspcc-dev/neo-go/pkg/services/notary.(*Notary).newTxCallbackLoop()
      /go/src/github.com/nspcc-dev/neo-go/pkg/services/notary/notary.go:369 +0x419

Previous write at 0x00c000298071 by goroutine 130:
  github.com/nspcc-dev/neo-go/pkg/core.TestNotary()
      /go/src/github.com/nspcc-dev/neo-go/pkg/core/notary_test.go:531 +0x22d7
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202

Goroutine 85 (running) created at:
  github.com/nspcc-dev/neo-go/pkg/services/notary.(*Notary).Run()
      /go/src/github.com/nspcc-dev/neo-go/pkg/services/notary/notary.go:130 +0x1a4

Goroutine 130 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1168 +0x5bb
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1439 +0xa6
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1123 +0x202
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1437 +0x612
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:1345 +0x3b3
  main.main()
      _testmain.go:239 +0x236
==================
```

File: pkg/services/notary/notary.go [hunk 1]
```diff
@@ -328,7 +322,72 @@ func (n *Notary) finalize(tx *transaction.Transaction) error {
...
+
+			if isMain {
+				r.isSent = true
+			} else {
+				for i := range r.fallbacks {
+					if r.fallbacks[i].Hash() == tx.tx.Hash() {
+						r.fallbacks = append(r.fallbacks[:i], r.fallbacks[i+1:]...)
+						break
+					}
+				}
```
Comment (AnnaShaleva) [inline pkg/services/notary/notary.go:384]: After this `for` loop we should check if all fallback transactions are sent `(len(r.fallbacks)==0)`. If so, then the request should be removed from the pool.

**Review by AnnaShaleva — COMMENTED** [review]
Something wrong with the test:
```
=== RUN   TestNotary
    logger.go:130: 2021-07-21T07:41:10.067Z	INFO	initial gas supply is not set or wrong, setting default value	{"InitialGASSupply": "52000000"}
    logger.go:130: 2021-07-21T07:41:10.067Z	INFO	P2PNotaryRequestPayloadPool size is not set or wrong, setting default value	{"P2PNotaryRequestPayloadPoolSize": 1000}
    logger.go:130: 2021-07-21T07:41:10.068Z	INFO	MaxBlockSize is not set or wrong, setting default value	{"MaxBlockSize": 262144}
    logger.go:130: 2021-07-21T07:41:10.068Z	INFO	MaxBlockSystemFee is not set or wrong, setting default value	{"MaxBlockSystemFee": 900000000000}
    logger.go:130: 2021-07-21T07:41:10.068Z	INFO	MaxTransactionsPerBlock is not set or wrong, using default value	{"MaxTransactionsPerBlock": 512}
    logger.go:130: 2021-07-21T07:41:10.068Z	INFO	MaxValidUntilBlockIncrement is not set or wrong, using default value	{"MaxValidUntilBlockIncrement": 5760}
    logger.go:130: 2021-07-21T07:41:10.070Z	INFO	no storage version found! creating genesis block
    logger.go:130: 2021-07-21T07:41:10.081Z	INFO	starting notary service
    logger.go:130: 2021-07-21T07:41:10.085Z	DEBUG	done processing headers	{"headerIndex": 1, "blockHeight": 0, "took": "1.083804ms"}
    logger.go:130: 2021-07-21T07:41:11.076Z	INFO	persisted to disk	{"blocks": 1, "keys": 135, "headerHeight": 1, "blockHeight": 1, "took": "133.721µs"}
    logger.go:130: 2021-07-21T07:41:11.841Z	DEBUG	done processing headers	{"headerIndex": 2, "blockHeight": 1, "took": "985.942µs"}
    notary_test.go:310: 
        	Error Trace:	notary_test.go:310
        	            				notary_test.go:461
        	Error:      	Expected nil, but got: &transaction.Transaction{Version:0x0, Nonce:0x33, SystemFee:100000000, NetworkFee:0, ValidUntilBlock:0x29, Script:[]uint8{0x40}, Attributes:[]transaction.Attribute{transaction.Attribute{Type:0xe2, Value:(*transaction.NotaryAssisted)(0xc000980408)}}, Signers:[]transaction.Signer{transaction.Signer{Account:util.Uint160{0x30, 0xc3, 0x82, 0x2c, 0x3d, 0xd6, 0x54, 0x4b, 0x26, 0xee, 0xd6, 0x55, 0xe2, 0x69, 0x27, 0xd, 0x62, 0x59, 0xbd, 0x6a}, Scopes:0x0, AllowedContracts:[]util.Uint160(nil), AllowedGroups:[]*keys.PublicKey(nil)}, transaction.Signer{Account:util.Uint160{0x3b, 0xec, 0x35, 0x31, 0x11, 0x9b, 0xba, 0xd7, 0x6d, 0xd0, 0x44, 0x92, 0xb, 0xd, 0xe6, 0xc3, 0x19, 0x4f, 0xe1, 0xc1}, Scopes:0x0, AllowedContracts:[]util.Uint160(nil), AllowedGroups:[]*keys.PublicKey(nil)}}, Scripts:[]transaction.Witness{transaction.Witness{InvocationScript:[]uint8{0xc, 0x40, 0x98, 0x18, 0xea, 0x96, 0x48, 0x40, 0x6d, 0xbc, 0x13, 0x52, 0x36, 0x56, 0x6b, 0x98, 0xf1, 0x29, 0xcd, 0x3f, 0x77, 0x23, 0xe7, 0x74, 0x81, 0x23, 0x1f, 0xb4, 0x39, 0xde, 0x59, 0xda, 0xdd, 0x84, 0x3d, 0x8f, 0x3d, 0x5f, 0xed, 0x47, 0xd1, 0x2e, 0x6e, 0x58, 0xeb, 0xf0, 0x9e, 0xb, 0x7a, 0x98, 0xe5, 0xe1, 0x5e, 0xae, 0x5e, 0xde, 0x1d, 0x5d, 0xd5, 0x9f, 0x49, 0xef, 0x92, 0x75, 0xd9, 0x90}, VerificationScript:[]uint8{0xc, 0x21, 0x2, 0xdf, 0x54, 0x3c, 0x3f, 0x77, 0xbd, 0xf8, 0xaf, 0xe6, 0x3, 0x40, 0xd9, 0x1d, 0xf9, 0x57, 0xeb, 0x2, 0xac, 0x4f, 0xc7, 0xe2, 0xb3, 0x69, 0xb6, 0xa6, 0x27, 0x4f, 0x42, 0x6, 0xb, 0x8e, 0x46, 0x41, 0x56, 0xe7, 0xb3, 0x27}}, transaction.Witness{InvocationScript:[]uint8{0xc, 0x40, 0xa0, 0xbf, 0xa3, 0xcd, 0x9e, 0x29, 0xf6, 0xa5, 0xe7, 0xc, 0xa0, 0xfc, 0xaa, 0x23, 0x23, 0xc0, 0xeb, 0x18, 0x21, 0xa7, 0xa7, 0xd6, 0xa3, 0x2b, 0xef, 0x57, 0x4d, 0x0, 0x7e, 0xbc, 0x7, 0x4e, 0x8c, 0x2a, 0x38, 0x10, 0x3d, 0x4d, 0x79, 0xb9, 0xc1, 0xb9, 0x50, 0xbc, 0xe6, 0x20, 0x61, 0x9f, 0x8b, 0xa, 0x47, 0xb7, 0xfc, 0x44, 0x5b, 0x38, 0xc2, 0x47, 0xbd, 0x4e, 0x1, 0x20, 0x80, 0x58}, VerificationScript:[]uint8{}}}, size:251, hash:util.Uint256{0x0, 0xf7, 0xca, 0xf2, 0x6f, 0x79, 0x87, 0x17, 0xf6, 0x60, 0x39, 0x3f, 0x37, 0xee, 0xe2, 0xba, 0xb7, 0xe1, 0xdf, 0xb9, 0x27, 0xd5, 0xbe, 0xfd, 0x1e, 0x6c, 0xb7, 0xaf, 0x50, 0x5d, 0xc, 0x4a}, Trimmed:false}
        	Test:       	TestNotary
        	Messages:   	main transaction shouldn't be completed: sent 1 out of 1 requests
    logger.go:130: 2021-07-21T07:41:11.842Z	INFO	persisted to disk	{"blocks": 1, "keys": 20, "headerHeight": 2, "blockHeight": 2, "took": "44.537µs"}
--- FAIL: TestNotary (1.78s)
```

File: pkg/core/notary_test.go [hunk 2]
```diff
@@ -91,6 +101,22 @@ func TestNotary(t *testing.T) {
...
+			return completedTxes[h]
+		}
+
+		var completedTx *transaction.Transaction
+		require.Eventually(t, func() bool {
+			mtx.RLock()
+			defer mtx.RUnlock()
+			completedTx = completedTxes[h]
+			return completedTx != nil
+		}, time.Second*3, time.Millisecond*50, errors.New("main transaction expected to be completed"))
```
Comment (AnnaShaleva) [inline pkg/core/notary_test.go:117]: I'm a bit worried about this test, it's still failing from time to time:
```
2021-07-21T13:30:09.1623849Z     logger.go:130: 2021-07-21T13:29:54.400Z	ERROR	new transaction callback finished with error	{"error": "error while finalizing transaction"}
2021-07-21T13:30:09.1625612Z     logger.go:130: 2021-07-21T13:29:54.760Z	INFO	persisted to disk	{"blocks": 43, "keys": 554, "headerHeight": 93, "blockHeight": 93, "took": "273.203µs"}
2021-07-21T13:30:09.1626622Z     notary_test.go:112: 
2021-07-21T13:30:09.1627098Z         	Error Trace:	notary_test.go:112
2021-07-21T13:30:09.1627688Z         	            				notary_test.go:341
2021-07-21T13:30:09.1628275Z         	            				notary_test.go:574
2021-07-21T13:30:09.1628767Z         	Error:      	Condition never satisfied
2021-07-21T13:30:09.1629373Z         	Test:       	TestNotary
2021-07-21T13:30:09.1630090Z         	Messages:   	main transaction expected to be completed
2021-07-21T13:30:09.1630937Z --- FAIL: TestNotary (5.66s)
2021-07-21T13:30:09.1645776Z === RUN   TestCreateResponseTx
```
Although code looks good to me and the problem is solved.

Comment (fyrchik) [reply pkg/core/notary_test.go]: I wasn't able to reproduce it locally, but increasing wait time seems to help. Probably it is related to built-time quotas on CI.

**Review by AnnaShaleva — APPROVED** [review]

**Review by roman-khimov — CHANGES_REQUESTED** [review]
I think with asynchronous `newTxCallbackLoop` we need to maintain some counter in `onTransaction`. We know exactly how many requests to expect (and we add blocks manually, so it can be precisely controlled), so we can `require.Eventually` for this counter to get to some value and then check/do more things.

I don't want randomly failing tests, so it's important.

**Comment by fyrchik** [comment]
Found and fixed another deadlock.

**Review by roman-khimov — APPROVED** [review]

**Review by AnnaShaleva — APPROVED** [review]

**Pull request merged**